node/src/components/consensus/era_supervisor/tests.rs

```diff
@@ -0,0 +1,120 @@
+use anyhow::Error;
+use casper_execution_engine::{core::engine_state::genesis::GenesisAccount, shared::motes::Motes};
+use derive_more::From;
+use prometheus::Registry;
+
+use super::*;
+use crate::{
+    components::{
+        consensus::{
+            tests::mock_proto::{Message, MockProto, NodeId},
+            Config,
+        },
+        storage::Storage,
+        Component,
+    },
+    crypto::asymmetric_key::{PublicKey, SecretKey},
+    effect::{
+        announcements::ConsensusAnnouncement,
+        requests::{
+            BlockExecutorRequest, BlockProposerRequest, BlockValidationRequest,
+            ContractRuntimeRequest, NetworkRequest, StorageRequest,
+        },
+        EffectBuilder,
+    },
+    protocol,
+    reactor::{EventQueueHandle, QueueKind, Scheduler},
+    testing::TestRng,
+    utils::{self, External, Loadable},
+};
+
+type ClMessage = Message<ClContext>;
+
+#[derive(Debug, From)]
+enum Event {
+    #[from]
+    Consensus(super::Event<NodeId>),
+    #[from]
+    Network(NetworkRequest<NodeId, protocol::Message>),
+    #[from]
+    BlockProposer(BlockProposerRequest),
+    #[from]
+    ConsensusAnnouncement(ConsensusAnnouncement),
+    #[from]
+    BlockExecutor(BlockExecutorRequest),
+    #[from]
+    BlockValidator(BlockValidationRequest<ProtoBlock, NodeId>),
+    #[from]
+    Storage(StorageRequest<Storage>),
+    #[from]
+    ContractRuntime(ContractRuntimeRequest),
+}
+
+/// Loads the local chainspec and overrides timestamp and genesis account with the given stakes.
+fn new_test_chainspec(stakes: Vec<(PublicKey, u64)>) -> Chainspec {
+    let mut chainspec = Chainspec::from_resources("local/chainspec.toml");
+    chainspec.genesis.accounts = stakes
+        .into_iter()
+        .map(|(pk, stake)| {
+            let motes = Motes::new(stake.into());
+            GenesisAccount::new(pk.into(), pk.to_account_hash(), motes, motes)
+        })
+        .collect();
+    chainspec.genesis.timestamp = Timestamp::now() + 1000.into();
+    chainspec.genesis.highway_config.genesis_era_start_timestamp = chainspec.genesis.timestamp;
+    chainspec
+}
+
+#[test]
+fn era_supervisor() -> Result<(), Error> {
```

Comment:
My attempted description of what is going on in this test:

1.  First, we construct an `EraSupervisor` corresponding to a node belonging to `alice`.  This is called `es` below.
2. Then, we simulate a validator `bob` sending a block and processing it with `es`.  This is done with the a `mock_proto::Message<ClContext>::BlockByOtherValidator` message
3. Finally, we simulate the block being finalized with a `mock_proto::Message<ClContext>::FinalizeBlock` message.

I understand there is further scope we'd like for this test, such as simulating a reactor to process effects emitted by the `EraSupervisor` after a block is finalized.  What else should we be testing here?

Reply:
The era supervisor is currently missing any unit tests at all, so: everything! Ideally I'd like to have unit tests for every non-trivial piece of logic in the era supervisor.
I'd turn this test case into the "good" case: Receive, propose and finalize a few blocks, and verify that we progress to the next era and initialize that correctly.
[NDRS-512](https://casperlabs.atlassian.net/browse/NDRS-512) will be specifically about testing that if you equivocate at the end of an era, you get slashed in the next.

Reply:
Can you add a description of what you want to test here? I don't like how Rust represents tests – with only a method name. It is not very descriptive.